### PR TITLE
add some ops

### PR DESCRIPTION
**Pull request by yeliang2258**
1. Add pool2d
2. scale、assign_value、depthwise_conv2d、softmax、transpose2、batch_norm、dropout、unsqueeze2、concat、split、flatten_contiguous_range、matmul、matmul_v2、clip

### CODE REVIEW

File: paddle2onnx/mapper/activation.cc [hunk 1]
```diff
@@ -218,4 +219,56 @@ void GeluMapper::Opset9(OnnxHelper* helper) {
...
   }
 }
+
+void SoftMaxMapper::Opset7(OnnxHelper* helper) {
+  auto op = parser_->GetOpDesc(block_idx_, op_idx_);
+  std::vector<TensorInfo> input_info =
+      parser_->GetOpInput(block_idx_, op_idx_, "X");
+  std::vector<TensorInfo> output_info =
+      parser_->GetOpOutput(block_idx_, op_idx_, "Out");
+  int64_t axis;
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/activation.cc]: 这里的获取移到Init里面吧，统一Attribute都在类的私有变量里声明

File: paddle2onnx/mapper/activation.cc [hunk 2]
```diff
@@ -218,4 +219,56 @@ void GeluMapper::Opset9(OnnxHelper* helper) {
...
+  }
+  if (axis == output_info[0].Rank() - 1) {
+    auto node = helper->MakeNode("Softmax", {input_info[0].name},
+                                 {output_info[0].name});
+    AddAttribute(node, "axis", axis);
+  } else {
+    std::vector<int64_t> perm;
+    for (auto i = 0; i < output_info[0].Rank(); i++) {
+      perm.push_back(i);
+    }
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/activation.cc]: auto perm = data_helper.h::Arange(0, output_info[0].Rank())

File: paddle2onnx/mapper/activation.h [hunk 3]
```diff
@@ -147,4 +147,13 @@ class GeluMapper : public Mapper {
...
 };
+
+class SoftMaxMapper : public Mapper {
+ public:
+  SoftMaxMapper(const PaddleParser& p, int64_t block_id, int64_t op_id)
+      : Mapper(p, block_id, op_id) {}
+
+  void Opset7(OnnxHelper* helper);
+  void Opset13(OnnxHelper* helper);
+};
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/activation.h:168]: axis变量声明在这

File: paddle2onnx/mapper/exporter.cc [hunk 4]
```diff
@@ -193,12 +193,14 @@ bool ModelExporter::CheckIfOpSupported(const PaddleParser& parser,
...
   int32_t max_opset = -1;
   bool exportable = true;
+  int valid_op_num = 0;
   for (auto i = 0; i < parser.NumOfBlocks(); ++i) {
     for (auto j = 0; j < parser.NumOfOps(i); ++j) {
       auto op = parser.GetOpDesc(i, j);
       if (op.type() == "feed" || op.type() == "fetch") {
         continue;
       }
+      valid_op_num += 1;
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/exporter.cc:210]: 为什么加这个逻辑？

File: paddle2onnx/mapper/nn/batch_norm.cc [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      parser_->GetOpInput(block_idx_, op_idx_, "Variance");
+  std::vector<TensorInfo> output_info =
+      parser_->GetOpOutput(block_idx_, op_idx_, "Y");
+
+  auto node = helper->MakeNode(
+      "BatchNormalization",
+      {input_info[0].name, scale_info[0].name, bias_info[0].name,
+       mean_info[0].name, variance_info[0].name},
+      {output_info[0].name});
+  if (export_opset_version_ < 9) {
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/batch_norm.cc]: opset_version统一从helper中获取吧，增加`ONNXHelper::GetOpserVersion()`接口
后面不再维护多个opset_version字段

File: paddle2onnx/mapper/nn/dropout.cc [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+REGISTER_MAPPER(dropout, DropoutMapper)
+
+int32_t DropoutMapper::GetMinOpset(bool verbose) {
+  if (dropout_implementation_ != "downgrade_in_infer" &&
+      dropout_implementation_ != "upscale_in_train") {
+    if (verbose) {
+      auto op = parser_->GetOpDesc(block_idx_, op_idx_);
+      std::cerr << "dropout_implementation: " << dropout_implementation_
+                << " is not support in op " << op.type() << "." << std::endl;
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/dropout.cc]: 已经知道是dropout了，为何还需要再获取op.type

File: paddle2onnx/mapper/nn/scale.cc [hunk 7]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
     // bias is 0
-    auto onnx_dtype = GetOnnxDtype(input_info[0].dtype);
-    auto bias_node = helper->MakeConstant({1}, onnx_dtype, bias_);
-    auto scale_node = helper->MakeConstant({1}, onnx_dtype, scale_);
+    int32_t data_type = input_info[0].dtype;
+    std::string output;
+    std::string cast_node;
+    if (input_info[0].dtype == P2ODataType::INT64 ||
+        input_info[0].dtype == P2ODataType::INT32 ||
+        input_info[0].dtype == P2ODataType::INT16) {
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc:44]: 这里是否是`if input_info[0].dtype != P2ODataType::FP32`  或者`input_info[0].dtype != FP32 && input_info[0].dtype != FP64`

File: paddle2onnx/mapper/nn/scale.cc [hunk 8]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
     // TODO(yeliang2258): we could add a pass to eleminate all the identity op
     helper->MakeNode("Identity", {input_info[0].name}, {output_info[0].name});
   } else {
     // TODO(yeliang2258): we could add a pass to eleminate the scale is 1 or
     // bias is 0
-    auto onnx_dtype = GetOnnxDtype(input_info[0].dtype);
-    auto bias_node = helper->MakeConstant({1}, onnx_dtype, bias_);
-    auto scale_node = helper->MakeConstant({1}, onnx_dtype, scale_);
+    int32_t data_type = input_info[0].dtype;
+    std::string output;
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc]: 看了下整体逻辑，output的作用是用来判断最终是否需要将结果cast回去
改成更直接的flag吧，逻辑也可以精简些
```
    bool is_int_input = false;
    std::string float_input_name = input_info[0].name;
    if (input_info[0].dtype == P2ODataType::INT64 ||
        input_info[0].dtype == P2ODataType::INT32 ||
        input_info[0].dtype == P2ODataType::INT16) {
      is_int_input = true;
      std::cerr << " Int type input may bring calculation diff in op "
                << op.type() << "." << std::endl;
      cast_node = helper->AutoCast(input_info[0].name, input_info[0].dtype,
                                   P2ODataType::FP32);
      float_input_name = cast_node->output(0);
    } 
```

File: paddle2onnx/mapper/nn/scale.cc [hunk 9]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
+      output = "";
+    } else {
+      output = output_info[0].name;
+      data_type = input_info[0].dtype;
+      cast_node = input_info[0].name;
+    }
+
+    std::string scale_node;
+    if (has_scale_tensor) {
+      std::vector<TensorInfo> scaletensor_info =
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc]: 变量命名使用下划线连接

File: paddle2onnx/mapper/nn/scale.cc [hunk 10]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
+          parser_->GetOpInput(block_idx_, op_idx_, "ScaleTensor");
+      scale_node = helper->AutoCast(scaletensor_info[0].name,
+                                    scaletensor_info[0].dtype, data_type);
+    } else {
+      std::vector<float> scale_vector = {scale_};
+      scale_node = helper->Constant(GetOnnxDtype(data_type), scale_vector);
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc]: 这里可直接写成`scale_node = helper->Constant({1}, GetOnnxDtype(input_info[0].dtype), scale_);`

File: paddle2onnx/mapper/nn/scale.cc [hunk 11]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
+    }
+
+    std::vector<float> bias_vector = {bias_};
+    std::string bias_node =
+        helper->Constant(GetOnnxDtype(data_type), bias_vector);
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc]: 同上

File: paddle2onnx/mapper/nn/scale.cc [hunk 12]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
+    std::string node2;
     if (bias_after_scale_) {
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc:67]: 代码中尽量减少这种未初始化的变量使用，在下面的代码逻辑中，有一个分支并没对这个变量进行操作，那这个变量的处理就完全取决于开发者对逻辑的信息了

File: paddle2onnx/mapper/nn/scale.cc [hunk 13]
```diff
@@ -19,32 +19,79 @@ namespace paddle2onnx {
...
-      auto mul_node =
-          helper->MakeNode("Mul", {input_info[0].name, scale_node->output(0)});
-      helper->MakeNode("Add", {mul_node->output(0), bias_node->output(0)},
-                       {output_info[0].name});
+      auto mul_node = helper->MakeNode("Mul", {cast_node, scale_node});
+      if (output.size() > 0) {
+        helper->MakeNode("Add", {mul_node->output(0), bias_node}, {output});
+      } else {
+        node2 = helper->MakeNode("Add", {mul_node->output(0), bias_node})
+                    ->output(0);
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/nn/scale.cc]: 可以将下面的Cast逻辑移到这个else里面来，移除上面的node2。 整体代码行数还更精简

File: paddle2onnx/mapper/onnx_helper.cc [hunk 14]
```diff
@@ -186,6 +191,12 @@ std::string OnnxHelper::AutoCast(const std::string& input,
   return cast_node->output(0);
 }
 
+std::shared_ptr<ONNX_NAMESPACE::NodeProto> OnnxHelper::MakeConstant(
+    const std::shared_ptr<ONNX_NAMESPACE::NodeProto>& node) {
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/onnx_helper.cc]: 这个函数的作用是什么

File: paddle2onnx/mapper/onnx_helper.cc [hunk 15]
```diff
@@ -364,4 +374,28 @@ std::vector<std::string> OnnxHelper::Split(const std::string& input,
...
+  Assert(input_info.size() > 0,
+         "OnnxHelper::DtypeAlignment requires the size of input info > 0.");
+  std::vector<int32_t> input_dtypes;
+  input_dtypes.reserve(input_info.size());
+  for (auto i = 0; i < input_info.size(); i++) {
+    input_dtypes.push_back(input_info[i].dtype);
+  }
+  int32_t max_index = -1;
+  for (auto i : input_dtypes) {
+    if (i > max_index) max_index = i;
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/onnx_helper.cc]: 我们的代码中，所有的条件分支都包在花括号内，统一风格

File: paddle2onnx/mapper/onnx_helper.cc [hunk 16]
```diff
@@ -364,4 +374,28 @@ std::vector<std::string> OnnxHelper::Split(const std::string& input,
   return Split(input, outputs, split, axis);
 }
 
+std::vector<std::string> OnnxHelper::DtypeAlignment(
+    const std::vector<TensorInfo>& input_info, int32_t& out_dtype) {
+  Assert(input_info.size() > 0,
+         "OnnxHelper::DtypeAlignment requires the size of input info > 0.");
+  std::vector<int32_t> input_dtypes;
+  input_dtypes.reserve(input_info.size());
+  for (auto i = 0; i < input_info.size(); i++) {
```
Comment (jiangjiajun) [inline paddle2onnx/mapper/onnx_helper.cc]: for循环也同样改成++i，统一风格吧

**Review by jiangjiajun — CHANGES_REQUESTED** [review]

File: paddle2onnx/mapper/activation.cc (hunk 1 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/activation.cc]: done

File: paddle2onnx/mapper/activation.cc (hunk 2 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/activation.cc]: done

File: paddle2onnx/mapper/activation.h (hunk 3 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/activation.h:168]: done

File: paddle2onnx/mapper/exporter.cc (hunk 4 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/exporter.cc:210]: 因为有可能有输入，有输出，但是没有OP

File: paddle2onnx/mapper/nn/batch_norm.cc (hunk 5 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/batch_norm.cc]: done

File: paddle2onnx/mapper/nn/dropout.cc (hunk 6 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/dropout.cc]: done

File: paddle2onnx/mapper/nn/scale.cc (hunk 7 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc:44]: 这是是如果为int类型则转换为float进行计算，pf32和pf64则按照原来的类型进行计算

File: paddle2onnx/mapper/nn/scale.cc (hunk 8 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc]: done

File: paddle2onnx/mapper/nn/scale.cc (hunk 9 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc]: done

File: paddle2onnx/mapper/nn/scale.cc (hunk 10 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc]: done

File: paddle2onnx/mapper/nn/scale.cc (hunk 11 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc]: done

File: paddle2onnx/mapper/nn/scale.cc (hunk 12 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc:67]: done

File: paddle2onnx/mapper/nn/scale.cc (hunk 13 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/nn/scale.cc]: done

File: paddle2onnx/mapper/onnx_helper.cc (hunk 14 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/onnx_helper.cc]: 已经修改，为nodes增加一个weight构造的const结点

File: paddle2onnx/mapper/onnx_helper.cc (hunk 15 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/onnx_helper.cc]: done

File: paddle2onnx/mapper/onnx_helper.cc (hunk 16 above)
Comment (yeliang2258) [reply paddle2onnx/mapper/onnx_helper.cc]: done

**Comment by jiangjiajun** [comment]
这个PR可以Close了吗？

**Comment by yeliang2258** [comment]
你不要管我这个PR，这个我对照来看的
